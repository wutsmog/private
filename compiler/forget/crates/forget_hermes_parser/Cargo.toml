[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }