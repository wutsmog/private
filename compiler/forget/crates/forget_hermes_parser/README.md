# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.