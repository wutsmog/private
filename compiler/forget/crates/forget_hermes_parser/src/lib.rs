mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
