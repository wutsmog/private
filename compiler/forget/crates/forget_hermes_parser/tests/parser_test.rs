use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
