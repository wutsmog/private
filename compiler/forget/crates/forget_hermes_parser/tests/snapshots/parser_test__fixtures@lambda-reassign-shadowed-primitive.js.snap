---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/lambda-reassign-shadowed-primitive.js
---
Input:
function Component() {
  const x = {};
  {
    let x = 56;
    const fn = function () {
      x = 42;
    };
    fn();
  }
  return x; // should return {}
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 9,
          "end": 18
        }
      },
      "params": [],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 31,
                    "end": 32
                  }
                },
                "init": {
                  "type": "ObjectExpression",
                  "properties": [],
                  "loc": null,
                  "range": {
                    "start": 35,
                    "end": 37
                  }
                },
                "loc": null,
                "range": {
                  "start": 31,
                  "end": 37
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 25,
              "end": 38
            }
          },
          {
            "type": "BlockStatement",
            "body": [
              {
                "type": "VariableDeclaration",
                "kind": "let",
                "declarations": [
                  {
                    "type": "VariableDeclarator",
                    "id": {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 51,
                        "end": 52
                      }
                    },
                    "init": {
                      "type": "NumericLiteral",
                      "value": 56.0,
                      "loc": null,
                      "range": {
                        "start": 55,
                        "end": 57
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 51,
                      "end": 57
                    }
                  }
                ],
                "loc": null,
                "range": {
                  "start": 47,
                  "end": 58
                }
              },
              {
                "type": "VariableDeclaration",
                "kind": "const",
                "declarations": [
                  {
                    "type": "VariableDeclarator",
                    "id": {
                      "type": "Identifier",
                      "name": "fn",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 69,
                        "end": 71
                      }
                    },
                    "init": {
                      "type": "FunctionExpression",
                      "id": null,
                      "params": [],
                      "body": {
                        "type": "BlockStatement",
                        "body": [
                          {
                            "type": "ExpressionStatement",
                            "expression": {
                              "type": "AssignmentExpression",
                              "operator": "=",
                              "left": {
                                "type": "Identifier",
                                "name": "x",
                                "typeAnnotation": null,
                                "loc": null,
                                "range": {
                                  "start": 94,
                                  "end": 95
                                }
                              },
                              "right": {
                                "type": "NumericLiteral",
                                "value": 42.0,
                                "loc": null,
                                "range": {
                                  "start": 98,
                                  "end": 100
                                }
                              },
                              "loc": null,
                              "range": {
                                "start": 94,
                                "end": 100
                              }
                            },
                            "directive": null,
                            "loc": null,
                            "range": {
                              "start": 94,
                              "end": 101
                            }
                          }
                        ],
                        "loc": null,
                        "range": {
                          "start": 86,
                          "end": 107
                        }
                      },
                      "generator": false,
                      "async": false,
                      "loc": null,
                      "range": {
                        "start": 74,
                        "end": 107
                      },
                      "loc": null,
                      "range": {
                        "start": 74,
                        "end": 107
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 69,
                      "end": 107
                    }
                  }
                ],
                "loc": null,
                "range": {
                  "start": 63,
                  "end": 108
                }
              },
              {
                "type": "ExpressionStatement",
                "expression": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "fn",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 113,
                      "end": 115
                    }
                  },
                  "arguments": [],
                  "loc": null,
                  "range": {
                    "start": 113,
                    "end": 117
                  }
                },
                "directive": null,
                "loc": null,
                "range": {
                  "start": 113,
                  "end": 118
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 41,
              "end": 122
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "Identifier",
              "name": "x",
              "typeAnnotation": null,
              "loc": null,
              "range": {
                "start": 132,
                "end": 133
              }
            },
            "loc": null,
            "range": {
              "start": 125,
              "end": 134
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 21,
          "end": 156
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 0,
        "end": 156
      },
      "loc": null,
      "range": {
        "start": 0,
        "end": 156
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 0,
    "end": 156
  }
}
